Lib/importlib/_bootstrap.py

```diff
@@ -614,30 +587,42 @@ def _exec(spec, module):
         if sys.modules.get(name) is not module:
             msg = 'module {!r} not in sys.modules'.format(name)
             raise ImportError(msg, name=name)
-        if spec.loader is None:
-            if spec.submodule_search_locations is None:
-                raise ImportError('missing loader', name=spec.name)
-            # namespace package
-            _init_module_attrs(spec, module, override=True)
-            return module
-        _init_module_attrs(spec, module, override=True)
-        if not hasattr(spec.loader, 'exec_module'):
-            # (issue19713) Once BuiltinImporter and ExtensionFileLoader
-            # have exec_module() implemented, we can add a deprecation
-            # warning here.
-            spec.loader.load_module(name)
-        else:
-            spec.loader.exec_module(module)
-    return sys.modules[name]
+        try:
+            if spec.loader is None:
+                if spec.submodule_search_locations is None:
+                    raise ImportError('missing loader', name=spec.name)
+                # namespace package
+                _init_module_attrs(spec, module, override=True)
+            else:
+                _init_module_attrs(spec, module, override=True)
+                if not hasattr(spec.loader, 'exec_module'):
+                    # (issue19713) Once BuiltinImporter and ExtensionFileLoader
+                    # have exec_module() implemented, we can add a deprecation
+                    # warning here.
+                    spec.loader.load_module(name)
+                else:
+                    spec.loader.exec_module(module)
+        finally:
+            module = sys.modules.pop(spec.name)
+            sys.modules[spec.name] = module
+    return module
 
 
 def _load_backward_compatible(spec):
     # (issue19713) Once BuiltinImporter and ExtensionFileLoader
     # have exec_module() implemented, we can add a deprecation
     # warning here.
-    spec.loader.load_module(spec.name)
+    try:
+        spec.loader.load_module(spec.name)
+    except:
+        if spec.name in sys.modules:
+            module = sys.modules.pop(spec.name)
```

Comment:
This pattern keeps coming up. Maybe add a function like `_move_to_sys_module_end(name)` or something?

Reply:
This pattern is short and can be written in a single line: `sys.modules[spec.name] = module = sys.modules.pop(spec.name)`. It is always added after `load_module()` and `exec_module()`.

Two occurrences in `_load_backward_compatible()` will be gone after removing support of this deprecated feature. Yet one in `_exec()` used by `importlib.reload()` and `_load_module_shim()`. Seems it will be removed too. And the single occurrence will be left in `_load_unlocked()`. I don't want to introduce a two-line function that will be used in a single place.

Reply:
OK, fair enough. Could you then please open issues to clean up that deprecated code? I don't know if an issue already exists to do this, else it's liable to just sit there for a while.

Reply:
Isn't this bpo-26131?

Reply:
Probably. 😄